### PR TITLE
Add Instant document JSON support

### LINKED ISSUE #37: Implement Instant JSON support
Instant JSON is our approach to bringing annotations into a modern format while keeping all important properties to make the Instant JSON spec work with PDF. We have already implemented an API to import and export Instant JSON for the other cross-platform wrappers [React Native](https://github.com/PSPDFKit/react-native/blob/122b8950d25e87edf8ecc41357fcfa678a84e12f/index.js#L279-L343) and [Cordova](https://github.com/PSPDFKit/Cordova-iOS#annotation-api).

## Public API
To keep consistency with the other APIs already implemented in the other wrappers we'll expose similar methods adapted for the PSPDFKit Dart plugin:

```dart 
Future<bool> applyInstantJson(String jsonValue) async
```
 
```dart 
Future <String> exportInstantJson() async
```

Both methods will rely on native calls to trigger the conversion to/from Instant JSON and will act asynchronously to guarantee a smooth UI.

## Android Host

PSPDFKit for Android uses [`DocumentJsonFormatter`](https://pspdfkit.com/api/android/reference/com/pspdfkit/document/formatters/DocumentJsonFormatter.html) to export and import changes made to annotations and form fields. [More info about Instant JSON for Android](https://pspdfkit.com/guides/android/current/importing-exporting/instant-json/)

## iOS Host
PSPDFKit for iOS uses [` -[PSPDFDocument generateInstantJSONFromDocumentProvider:error:]`](https://pspdfkit.com/api/ios/Classes/PSPDFDocument.html#/c:objc(cs)PSPDFDocument(im)generateInstantJSONFromDocumentProvider:error:) and [ -[PSPDFDocument applyInstantJSONFromDataProvider:toDocumentProvider:lenient:error:]](https://pspdfkit.com/api/ios/Classes/PSPDFDocument.html#/c:objc(cs)PSPDFDocument(im)applyInstantJSONFromDataProvider:toDocumentProvider:lenient:error:).
[More info about Instant JSON for iOS](https://pspdfkit.com/guides/ios/current/importing-exporting/instant-json/)

## Example

A new section for both platform will be added in our general example app to showcase how to import ~and export~ Instant Document JSON with the PSPDFKit Flutter plugin.

## Readme

Expand the main `README` file with a new section called Instant Document JSON where the new API is explained and the example is referenced.

### PR DESCRIPTION
**Pull request by simoarpe**
Resolves #37 

# Details
To keep consistency with the other APIs already implemented in the other wrappers this PR exposes similar methods adapted for the PSPDFKit Dart plugin:

```dart
Future<bool> applyInstantJson(String annotationsJson) async
Future <String> exportInstantJson() async
```

Both methods rely on native calls to trigger the conversion to/from Instant JSON and will act asynchronously to guarantee a smooth UI.

Among the other changes:

- Improved Android `Preconditions` class
- Updated example document with one containing more up-to-date content
- Better use of async/await concepts
- Added new section in the example app `Import Instant Document JSON`

instant-json

# Acceptance Criteria

- [x] Expose Dart APIs
- [x] Include new example
- [x] Android part implemented
- [x] iOS part implemented
- [x] Before merging retest all the examples to make sure that they work on both Android and iOS.

### CODE REVIEW

**Comment by aksh1t** [comment]
I added the iOS implementation for the instant JSON methods and tested the examples too.

File: ios/Classes/PspdfkitPlugin.m [hunk 1]
```diff
@@ -37,6 +37,39 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
     } else if ([@"getFormFieldValue" isEqualToString:call.method]) {
         NSString *fullyQualifiedName = call.arguments[@"fullyQualifiedName"];
         result([self getFormFieldValueForFieldWithFullyQualifiedName:fullyQualifiedName]);
+    } else if ([@"applyInstantJson" isEqualToString:call.method]) {
+        NSString *annotationsJson = call.arguments[@"annotationsJson"];
+        if (annotationsJson == nil || annotationsJson.length <= 0) {
```
Comment (radazzouz) [inline ios/Classes/PspdfkitPlugin.m]: I think we could simplify the condition to:

```suggestion
        if (annotationsJson.length == 0) {
```

File: ios/Classes/PspdfkitPlugin.m [hunk 2]
```diff
@@ -37,6 +37,39 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
+    } else if ([@"exportInstantJson" isEqualToString:call.method]) {
+        PSPDFDocument *document = self.pdfViewController.document;
+        if (!document || !document.isValid) {
+            result([FlutterError errorWithCode:@"" message:@"PDF document not found or is invalid." details:nil]);
+            return;
+        }
+        NSError *error;
+        NSData *data = [document generateInstantJSONFromDocumentProvider:document.documentProviders.firstObject error:&error];
+        NSString *annotationsJson = [[NSString alloc]initWithData:data encoding:NSUTF8StringEncoding];
+        if (annotationsJson == nil || annotationsJson.length <= 0 || error) {
```
Comment (radazzouz) [inline ios/Classes/PspdfkitPlugin.m]: I am not very fond of checking if the error is not nil. I think that we should just check if `annotationsJson` is nil.

```suggestion
        if (annotationsJson == nil) {
```

File: ios/Classes/PspdfkitPlugin.m [hunk 3]
```diff
@@ -37,6 +37,39 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
+            result([FlutterError errorWithCode:@"" message:@"Error while exporting document Instant JSON." details:nil]);
```
Comment (radazzouz) [inline ios/Classes/PspdfkitPlugin.m]:
```suggestion
            result([FlutterError errorWithCode:error.code message:@"Error while exporting document Instant JSON." details:error.localizedDescription]);
```

**Comment by simoarpe** [comment]
> I think we should improve our example to illustrate how to import and export Instant JSON. We currently only showcase the import scenario. We should also showcase how to export.

> I think we should add custom buttons for importing and exporting Instant JSON payloads, similarly to what we do on React Native.

Having a different UI containing buttons and other elements it's not supported. It requires a native UI component, [like we do in React Native](https://github.com/PSPDFKit/react-native#native-ui-component).
So even if would be a nice addition for our examples, it's not something we can achieve right now as it requires a different component.

**Comment by radazzouz** [comment]
Makes sense! How about programmatically showing an alert after a few seconds with the export Instant JSON payload?

**Pull request merged**